modules/test/playwright/fixtures/featureFlagsTest.ts

```diff
@@ -0,0 +1,171 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {Page, mergeTests, test} from '@playwright/test';
+
+import {loginTest} from './loginTest';
+
+export type FeatureFlagsOptions = {
+	[key: string]: boolean;
+};
+
+export type FeatureFlags = FeatureFlag[];
+
+export interface FeatureFlag {
+	readonly enabled: boolean;
+	readonly key: string;
+}
+
+function featureFlagsTest(options: FeatureFlagsOptions) {
+	const fixtureImpl = test.extend<{
+		featureFlags: FeatureFlags;
+	}>({
+		featureFlags: [
+			async ({page}, use) => {
+
+				// Save original state of FFs
+
+				const originalFeatureFlags: FeatureFlagResult[] = [];
+
+				for (const key of Object.keys(options)) {
+					const result = await isEnabled(key, page);
+
+					originalFeatureFlags.push(result.featureFlag);
+					originalFeatureFlags.push(...result.dependentFeatureFlags);
+				}
+
+				try {
+
+					// Activate  FFs
+
+					for (const [key, enabled] of Object.entries(options)) {
+						await setEnabled(enabled, key, page);
+					}
+
+					// Run test
+
+					await use(
+						Object.entries(options).reduce(
+							(featureFlags: FeatureFlags, [key, enabled]) => {
+								featureFlags.push({
+									enabled,
+									key,
+								});
+
+								return featureFlags;
+							},
+							[]
+						)
+					);
+				}
+				finally {
+
+					// Restore original state of FFs
+
+					originalFeatureFlags.reverse();
+
+					for (const featureFlag of originalFeatureFlags) {
+						await setEnabled(
+							featureFlag.enabled,
+							featureFlag.key,
+							page
+						);
+					}
```

Comment:
I am not sure if we want to immediately restore original state. What if FF changes backend behavior, and tests depend on the changed behavior? I would assume FF recovery needs to be done only after all tests are completed.

Reply:
> What if FF changes backend behavior, and tests depend on the changed behavior?

Well, if FF changes behavior and it is needed for another test you have to use the fixture from other test. That's the intention of this fixture: eliminating side effects (from one tests to others).

> I would assume FF recovery needs to be done only after all tests are completed.

It is set and restored per test. So, in the same file, if there are 3 tests using it those 3 tests will turn the FF on on entering, then restoring it on exit.

I think the misunderstanding here is that you think the declaration  `featureFlagsTest(...)` is affecting the whole file but that is not true. What `mergeTests` does is composing several fixtures and it's after, when you invoke the returned `test`  method, that the fixture is really used.

In other words, you could have two different mergings in the same file, which clearly demonstrates what I'm tying to convey:


```typescript
const testWithFFa = mergeTests(loginTest, featureFlagsTest({"LPS-A": true}));
const testWithFFb = mergeTests(loginTest, featureFlagsTest({"LPS-B": true}));

testWithFFa('something happens' ...);

testWithFFa('other things happens' ...);
```

Reply:
One possible optimization to the current logic would be letting the FFs changed at the end of the tests, then change them only when entering the setup phase again.

However this optimization needs the state of the server to be stored somewhere so that we don't lose track of what it is. For one single worker that's definitely possible if we promote the `originalFeatureFlags` variable to global module scope. 

However, for several workers that setup wouldn't work because we would need to distribute that information among the workers which would be difficult (even impossible).

It's true that with several workers even the current code fails because two workers may collide when setting/restoring FFs, but at least the current code seems more stable (to me). I prefer to start slow (conservative) :shrug:

Reply:
No no, I didn't mean about interaction between tests. Assume we have only one test. What I meant is that test may depend on what FF does in backend. FFs don't only toggle parts of UI, they change Java logic, enable endpoints etc. A fixture resolves before test, and the current solution in PR immediately restores it. If I'm understanding this correctly, this means the state in the tests would be UI with overridden FFs, but the backed with unchanged FFs. If tests depend on changed backend logic, they will fail.

Reply:
For the record, I talked to @markocikos by DM and the misunderstanding comes from the fact that he wasn't aware that the invocation of `use()` [inside the fixture](https://github.com/liferay-frontend/liferay-portal/pull/3955/files#diff-ab0aab68b4c280e9bdbf15330c8303a7459aec9c465862b2a90f09d3eab5f349R81) is in fact executing the test. 

Knowing that everything makes sense :+1: .